subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/loader/DefaultToolingImplementationLoader.java

```diff
@@ -97,20 +98,24 @@ public ConsumerConnection create(Distribution distribution, ProgressLoggerFactor
             } else if (connection instanceof BuildActionRunner) {
                 adaptedConnection = new BuildActionRunnerBackedConsumerConnection(connection, modelMapping, adapter);
             } else {
-                return new UnsupportedOlderVersionConnection(connection, adapter);
+                return new VersionControlConsumerConnection(connection, adapter);
             }
             adaptedConnection.configure(connectionParameters);
-            if (!adaptedConnection.getVersionDetails().supportsCancellation()) {
-                return new ParameterValidatingConsumerConnection(adaptedConnection.getVersionDetails(), new NonCancellableConsumerConnectionAdapter(adaptedConnection));
-            }
-            return new ParameterValidatingConsumerConnection(adaptedConnection.getVersionDetails(), adaptedConnection);
+
+            return createConnection(adaptedConnection);
         } catch (UnsupportedVersionException e) {
             throw e;
         } catch (Throwable t) {
             throw new GradleConnectionException(String.format("Could not create an instance of Tooling API implementation using the specified %s.", distribution.getDisplayName()), t);
         }
     }
 
+    private ConsumerConnection createConnection(AbstractConsumerConnection adaptedConnection) {
+        VersionDetails versionDetails = adaptedConnection.getVersionDetails();
+        ConsumerConnection delegate = versionDetails.supportsCancellation() ? adaptedConnection : new NonCancellableConsumerConnectionAdapter(adaptedConnection);
+        return new VersionControlConsumerConnection(new ParameterValidatingConsumerConnection(versionDetails, delegate), versionDetails);
+    }
+
```

Comment:
I slightly refactor this, providing a unified `VersionControlConsumerConnection`.